src/canister_sig_util/src/lib.rs

```diff
@@ -1 +1,398 @@
+use candid::Principal;
+use ic_certification::{Certificate, Delegation, HashTree, LookupResult};
+use ic_certified_map::Hash;
+use ic_verify_bls_signature::verify_bls_signature;
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use serde_bytes::ByteBuf;
+use sha2::{Digest, Sha256};
+use std::fmt::{Display, Formatter};
+use std::sync::RwLock;
+
 pub mod signature_map;
+
+pub const IC_STATE_ROOT_DOMAIN_SEPARATOR: &[u8; 14] = b"\x0Dic-state-root";
+pub const IC_ROOT_PK_DER_PREFIX: &[u8; 37] = b"\x30\x81\x82\x30\x1d\x06\x0d\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x01\x02\x01\x06\x0c\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x02\x01\x03\x61\x00";
+pub const IC_ROOT_PK_DER: &[u8; 133] = b"\x30\x81\x82\x30\x1d\x06\x0d\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x01\x02\x01\x06\x0c\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x02\x01\x03\x61\x00\x81\x4c\x0e\x6e\xc7\x1f\xab\x58\x3b\x08\xbd\x81\x37\x3c\x25\x5c\x3c\x37\x1b\x2e\x84\x86\x3c\x98\xa4\xf1\xe0\x8b\x74\x23\x5d\x14\xfb\x5d\x9c\x0c\xd5\x46\xd9\x68\x5f\x91\x3a\x0c\x0b\x2c\xc5\x34\x15\x83\xbf\x4b\x43\x92\xe4\x67\xdb\x96\xd6\x5b\x9b\xb4\xcb\x71\x71\x12\xf8\x47\x2e\x0d\x5a\x4d\x14\x50\x5f\xfd\x74\x84\xb0\x12\x91\x09\x1c\x5f\x87\xb9\x88\x83\x46\x3f\x98\x09\x1a\x0b\xaa\xae";
+pub const IC_ROOT_PK_LENGTH: usize = 96;
+
+pub const CANISTER_SIG_PK_DER_PREFIX_LENGTH: usize = 19;
+// Canister signatures' public key OID is 1.3.6.1.4.1.56387.1.2,
+// cf. https://internetcomputer.org/docs/current/references/ic-interface-spec#canister-signatures
+pub const CANISTER_SIG_PK_DER_OID: &[u8; 14] =
+    b"\x30\x0C\x06\x0A\x2B\x06\x01\x04\x01\x83\xB8\x43\x01\x02";
+
+lazy_static! {
+    /// The root public key used when verifying canister signatures.
+    static ref IC_ROOT_PUBLIC_KEY: RwLock<Vec<u8>> = RwLock::new(
+        extract_ic_root_key_from_der(IC_ROOT_PK_DER).expect("Failed decoding IC root key.")
+    );
+}
+
+/// Resets the root public key to the given value, should be used for testing only.
+pub fn set_ic_root_public_key_for_testing(pk_der: Vec<u8>) {
```

Comment:
I don't like this:
* it makes the tests stateful, hence the `#[serial]` annotation
* so far we have been careful to only have state in a single place (so that we can easily find all actions that mutate state). Scattering global mutable state in a library makes this much harder to track.

I think it would be nicer to have the following:
* private fn that takes the root key as an argument (the enclosed `test` module has access to that)
* a public wrapper that calls the internal function above, but supplies the static IC root key.

Reply:
That's a good idea, although comes with a drawback that the tests do test the private function, rather than the public one. Still, I agree the public wrapper is simple enough, so done.